src/functional/otf/compilation/compiler.py

```diff
@@ -0,0 +1,91 @@
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+from __future__ import annotations
+
+import dataclasses
+import pathlib
+from typing import Protocol, TypeVar
+
+from functional.otf import languages, stages, step_types, workflow
+from functional.otf.compilation import build_data, cache, importer
+from functional.otf.step_types import LS, SrcL, TgtL
+
+
+SourceLanguageType = TypeVar("SourceLanguageType", bound=languages.LanguageTag)
+LanguageSettingsType = TypeVar("LanguageSettingsType", bound=languages.LanguageSettings)
+T = TypeVar("T")
+
+
+def is_compiled(data: build_data.BuildData) -> bool:
+    return data.status >= build_data.BuildStatus.COMPILED
+
+
+def module_exists(data: build_data.BuildData, src_dir: pathlib.Path) -> bool:
+    return (src_dir / data.module).exists()
+
+
+class BuildSystemProjectGenerator(Protocol[SrcL, LS, TgtL]):
+    def __call__(
+        self,
+        source: stages.CompilableSource[SrcL, LS, TgtL],
+        cache_strategy: cache.Strategy,
+    ) -> stages.BuildSystemProject[SrcL, LS, TgtL]:
+        ...
+
+
+@dataclasses.dataclass(frozen=True)
+class Compiler(
+    step_types.CompilationStep[SourceLanguageType, LanguageSettingsType, languages.Python]
+):
+    cache_strategy: cache.Strategy
+    builder_factory: BuildSystemProjectGenerator[
+        SourceLanguageType, LanguageSettingsType, languages.Python
+    ]
+    force_recompile: bool = False
+    """Use any build system (via configured factory) to compile a GT4Py program to a ``functional.otf.stages.CompiledProgram``."""
+
+    def __call__(
+        self,
+        inp: stages.CompilableSource[SourceLanguageType, LanguageSettingsType, languages.Python],
+    ) -> stages.CompiledProgram:
+        src_dir = cache.get_cache_folder(inp, self.cache_strategy)
+
+        data = build_data.read_data(src_dir)
+
+        if not data or not is_compiled(data) or self.force_recompile:
+            self.builder_factory(inp, self.cache_strategy).build()
+
+        new_data = build_data.read_data(src_dir)
+
+        if not new_data or not is_compiled(new_data) or not module_exists(new_data, src_dir):
+            raise CompilerError(
+                "On-the-fly compilation unsuccessful for {inp.source_module.entry_point.name}!"
+            )
+
+        return getattr(
+            importer.import_from_path(src_dir / new_data.module), new_data.entry_point_name
+        )
+
+    def chain(
+        self, step: workflow.StepProtocol[stages.CompiledProgram, T]
+    ) -> workflow.Workflow[
+        stages.CompilableSource[SourceLanguageType, LanguageSettingsType, languages.Python],
+        stages.CompiledProgram,
+        T,
+    ]:
+        return workflow.Workflow(first=self, second=step)
+
+
+class CompilerError(Exception):
```

Comment:
Assuming it does not match a builtin exception (there is only a handful, so I encourage you to look at them and check if you find a matching one, I didn't): You are basically saying it should be called `CompilationError` because that tells the user the compilation didn't work, which is more helpful than to know that something in the compiler component went wrong.

Reply:
For compiler input validation it would make sense to derive from `ValueError`, as that is in line with the semantics of `ValueError`. However, not making it a custom subclas of `ValueError` is missing an opportunity to make debuggin easier. Making the error type for when the compiler fails a subclass of `ValueError` is not in line with it's semantics, that would be a `RuntimeError`. These changes are simple enough and deserve to be included.

Reply:
There is now a `CompilerInputError`, derived from `ValueError` as well as a `CompilationError`, derived from `RuntimeError`.